File: leetcode problems/Combination Sum.cpp
class Solution
{
public:
    void find(int id, vector<int> &candidates, int target, vector<vector<int>> &ans, vector<int> &temp)
    {
        if (id == candidates.size())
        {
            if (target == 0)
                ans.push_back(temp);
            return;
        }

        if (candidates[id] <= target)
        {
            temp.push_back(candidates[id]);
            find(id, candidates, target - candidates[id], ans, temp);
            temp.pop_back();
        }

        find(id + 1, candidates, target, ans, temp);
    }

    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {

        vector<vector<int>> ans;
        vector<int> temp;

        find(0, candidates, target, ans, temp);

        return ans;
    }
};

File: Java/CombinationSum.java
// this was the most common and highlt asked problem on FAANG Interviews from Leetcode
// Approach - BackTracking 

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {

        List<Integer> active = new ArrayList<>();
        List<List<Integer>> ans = new ArrayList<>();
        helper(0, target, candidates, active, ans);
        return ans;
    }
    static void helper (int index, int target, int[] candidates, List<Integer> active, List<List<Integer>> ans) {

        if (target == 0) {
            ans.add(new ArrayList<>(active));
            return;
        }

        if (index == candidates.length) {
            return;
        }

        if (candidates[index] <= target) {
            active.add(candidates[index]);
            helper (index, target - candidates[index], candidates, active, ans);
            active.remove(active.size() - 1);
        }
        helper (index + 1, target, candidates, active, ans);
    }
}
